### DISCUSSION TITLE
Command alerter formatting for Curl command

### DISCUSSION THREAD

**Discussion by firdausa-iprogrammer** [jertel/elastalert2#492] [Q&A] [answered]
Hi!

So sorry to bug you guys about this again, my company now wants to switch from Signal to Telegram.
I'm still able to use the alerter in the answer above, I have a minor error however -

When I was using signal-cli for alerts \n worked to get the text in a new line.
For Telegram I'm using curl instead of the alerter that's already present to achieve my goal of threshold based alerting.
My curl command -

curl -XPOST https://api.telegram.org/bot<botToken>/sendMessage?chat_id=<chat_id> -d "text=- Alert_level: L1(Devops)                                                                                                         
   - Alert_Type: Every 10 minutes 
  - APIName: <alert name>
  - etc
When I fire this same command from bash even with sh -c "" which I believe subprocess uses I get the output on Telegram perfectly fomatted I don't even need \n.

However when I trigger the alert from elastalert it sends all the data but with only about 2 character space
for eg.

- Alert_level: L1(Devops)  - Alert_Type: Every 10 minutes - APIName: <APIName>
How do I format my curl command to work like it works on it's own?

```
command_l1: 'curl -XPOST https://api.telegram.org/bot<botToken>/sendMessage?chat_id=<chat_id> -d "text=- Alert_level: L1-Devops \n
  - Alert_Type: Every 10 minutes \n
  - APIName: <APIName> \n
  - ErrorCode: <ErrorCode> \n
  - Alert start time: {starttime_ist} \n
  - Alert end time:   {endtime_ist} \n
  - Threshold: 5 %% in the Last 10 minutes \n
  - Current_Failure_%% in the Last 10 minutes: {percentage:.2f} %% \n
  - Error_Message: <Error_Message> \n
  - ErrorDescription: <ErrorDescription> \n
  - SOR: <SOR> \n
  - Action to be taken: \n
    1. Automated Notification Email sent to L1-Devops. \n
    2. SOR team must revert in next 2 hrs \n
    3. if revert not received then DevOps team contact SOR team for update \n
    4. Identify the issue & update on same in signal grp"'
```
Using ```alert_text_jinja```

I'm using a custom alerter that I've created in https://github.com/jertel/elastalert2/discussions/312#discussioncomment-1012110

Do let me know if any other data is needed.

**Comment by nsano-rururu** [comment]
The command handler supports the execution of curl commands, so if you look at the code and implement it yourself
https://github.com/jertel/elastalert2/blob/master/elastalert/alerters/command.py

**Reference information**

Problems with "command" alert
https://github.com/Yelp/elastalert/issues/592

**Reply by firdausa-iprogrammer** [reply]
@nsano-rururu my alerter directly calls subprocess.run, I haven't integrated the command alerter.

**Reply by nsano-rururu** [reply]
It doesn't matter to me because it's about thinking about what you do. Just refer to the command alert code. Do anything beyond that on your own.

**Reply by nsano-rururu** [reply]
https://gist.github.com/zacharyvoase/4067466

**Reply by nsano-rururu** [reply]
Search google for subprocess and curl as keywords.

**Reply by nsano-rururu** [reply]
https://www.fixes.pub/program/55514.html

**Reply by nsano-rururu** [reply]
It's not a place to hear how to write python code

**Reply by nsano-rururu** [reply]
https://stackoverflow.com/questions/26000336/execute-curl-command-within-a-python-script

**Reply by firdausa-iprogrammer** [reply]
@nsano-rururu,

I am not a developer and a complete newbie when it comes to Python, hence I wanted to know from the community whether I had made a mistake in my custom alerter.
I only mentioned the subprocess point because I thought it would help you experts in diagnosing the issue.

Please do forgive me if I broke any discussion rules by asking about this.

Thank for the suggestions anyways!
The last one makes a lot of sense to switch to request instead of running a shell.

**Comment by jertel** [comment, marked as answer]
I suggest printing the `signal` variable immediately before calling subprocess.run() to see if the newlines have been stripped already. That will help isolate whether the problem is with the yaml string encoding, or with the subprocess.run() command.

**Reply by firdausa-iprogrammer** [reply]
Great idea @jertel I'll check this point out as well.

I printed the signal variable like you suggested and the output came as ```\\n``` for the newlines
Does ```alert_text_jinja``` still take effect here?
I already tried ```{% autoescape false %}``` for the data section of the curl command but that didn't help either.
I'm still not sure whether Jinja is stripping the newlines or Python and the funny thing is when I used this same alerter and config for ```signal-cli``` it worked perfectly.

**Reply by jertel** [reply]
It's difficult to say what the problem from a distance. I would recommend trying to set the variable with some variations of \n, \r\n, etc to see which combination allows the newlines to go through.

**Reply by firdausa-iprogrammer** [reply]
Hi @jertel,

I solved it by tweaking a few things -

In the rule's command section I used yaml ```pipe(|)``` instead of ```""```

I also created a new alert that uses requests instead of subprocess, thanks to @nsano-rururu for pointing me in the right direction.
The issue was that since we were using the ```|``` we were getting a list so as an additional step I had to initialize an empty variable and in the message data section of the alert I used a ```.join()```.
Also switched to f-strings for more readability

The alert is below for anyone who faces a similar problem -

```
from elastalert.alerts import Alerter
import requests
from elastalert.util import resolve_string

class fidAlerter(Alerter):
  required_options = set(['command_l1', 'command_l2', 'command_l3', 'thresh_l1_max', 'thresh_l2', 'thresh_l2_max'])
  def alert(self, matches):
      for match in matches:
        pct = match['percentage']
        thresh_l1 = self.rule['max_percentage']
        thresh_l1_max = self.rule['thresh_l1_max']
        thresh_l2 = self.rule['thresh_l2']
        thresh_l2_max = self.rule['thresh_l2_max']
        if pct >= thresh_l1 and pct <= thresh_l1_max:
          self.rule['command_l1'] = [self.rule['command_l1']]
          payload = [resolve_string(command_l1, matches[0]) for command_l1 in self.rule['command_l1']]
          text = " "
          url = f'https://api.telegram.org/bot<botToken>/sendMessage?chat_id=<chat_id>&text={text.join(payload)}'
          res = requests.post(url)
          print(res.text)
        elif pct >= thresh_l2 and pct <= thresh_l2_max:
          self.rule['command_l2'] = [self.rule['command_l2']]
          payload = [resolve_string(command_l2, matches[0]) for command_l2 in self.rule['command_l2']]
          text = " "
          url = f'https://api.telegram.org/bot<botToken>/sendMessage?chat_id=<chat_id>&text={text.join(payload)}'
          res = requests.post(url)
          print(res.text)
        else:
          self.rule['command_l3'] = [self.rule['command_l3']]
          payload = [resolve_string(command_l3,matches[0]) for command_l3 in self.rule['command_l3']]
          text = " "
          url = f'https://api.telegram.org/bot<botToken>/sendMessage?chat_id=<chat_id>&text={text.join(payload)}'
          res = requests.post(url)
          print(res.text)

  def get_info(self):
    return {'type': 'Threshold Alerter'}
```

Thanks again to both of you for the support!!
If you feel I could optimise or use a different approach please feel free to let me know.
